pennylane/_qubit_device.py

```diff
@@ -94,7 +94,9 @@ class QubitDevice(Device):
     _reduce_sum = staticmethod(lambda array, axes: np.sum(array, axis=tuple(axes)))
```

Comment:
I know this was discussed previously, but does it make sense to make all these changes at the _qubit_device level? Many of these updates would be better suited to be focused entirely to `default.qubit` in my eyes, or even a separate bcast_device that serves the purpose of providing this functionality directly to its children.

Reply:
I see your point, I'm pretty undecided what's best here.
An option could indeed be a Mixin device class that basically overwrites exactly the methods changed in `_qubit_device.py` in this PR. In contrast to pushing all changes into `default_qubit.py`, this would allow devices to support broadcasting without having to inherit from DefaultQubit.
Then one could have an optional kwarg `broadcast` to the `qml.device` method. I would activate it by default if `"default.qubit"` or any of the interface-specific variants is used, and deactivate it by default for other devices.

Even more radically, one could additionally include all changes of this PR to `default_qubit.py` in a Mixin. However
1. If we include these changes in the mixin above, the `QubitDevice` and `DefaultQubit` abstraction levels would be mixed, probably making the Mixin useless for other devices than `DefaultQubit`. For example, if we really want `LightningQubit` to inherit from `QubitDevice` instead of `DefaultQubit` in the long run, the Mixin might not be applicable.
2. If we create both, a `QubitDeviceBroadcastingMixin` and a `DefaultQubitBroadcastingMixin`, it might become very confusing?

Reply:
One of the problems I see in the current design is that `default.qubit` is both the parent class of other devices (e.g. default.qubit.torch`) and implementing `autograd` specific features. In the long run, I think it is better to have another parent class (`QubitDeviceAutoDiff`?) derived from `QubitDevice` and has current devices supporting automatic differentiation as children. As all these devices naturally support broadcasting, I would expect this design choice also solves the issue here.

Reply:
Interesting. I must admit, this seems like a fundamental refactor of the device infrastructure. :sweat_smile:

Reply:
@chaeyeunpark we've been wanting to merge all the `default.qubit` subdevices back into `default.qubit` for a long time 😆

Reply:
@josh146 Interesting. That also sounds nice if we can make a better abstraction of all arrays (`autograd`, `torch`, `tf`, `jax`) that works the same regardless of the underlying library. Still, I think that as all existing tools do not support all functionalities we want (`autoray`?), we need to make our own abstraction (which requires lots of effort).

Reply:
I think we _could_ rewrite `default.qubit` to use autoray/`qml.math`, and then it would 'just work' with all autodiff frameworks for backprop.

My worry (I think @dwierichs confirmed this at one stage) that autoray is slower than having separate devices, due to the dispatch overhead.

Reply:
Does it? I thought it might be more involved as some of the functions we use is not precisely at the intersection of `tensorflow`, `pytorch`, and `jax` (I might be just wrong). In any case, it would be worth investigating a difficulty (or performance regression) in integrating all subdevices.

Reply:
I think there would be some details that differ for the interfaces, leading to a few methods probably needing their `..._tf` or `..._jax` version, but I believe most parts should be possible to collect using `qml.math`.
Indeed the performance question is open here. Maybe there is a smart way of not calling `get_interface` too often but somehow "caching" which interface is being used with the device?

I feel like all this discussion might belong somewhere else? :)

Reply:
Yes it probably does belong somewhere else :)

Something that has been on my mind for a while: I wonder if supporting a UX like the following would help negate a lot of the dispatch overhead:

```python
>>> qml.math.set_backend("tf")
>>> qml.math.sin(x)  # assumes TF since the backend is set
>>> qml.math.set_backend("auto")  # revert to dispatch
```